doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -216,50 +255,62 @@ might be affected by this change:
             func=broadcast_to,
             relevant_arguments=[array],
             args=(array,),
             kwargs=dict(shape=shape, subok=subok))
         if success:
             return value
 
-        ... # continue with the definition of broadcast_to
+        ...  # continue with the definition of broadcast_to
 
     def concatenate(arrays, axis=0, out=None)
-        success, value = do_array_function_dance(
+        success, value = try_array_function_override(
             func=concatenate,
-            relevant_arguments=[arrays, out],
+            relevant_arguments=itertools.chain(arrays, [out]),
             args=(arrays,),
             kwargs=dict(axis=axis, out=out))
         if success:
             return value
 
-        ... # continue with the definition of concatenate
+        ...  # continue with the definition of concatenate
 
 The list of objects passed to ``relevant_arguments`` are those that should
 be inspected for ``__array_function__`` implementations.
 
-Alternatively, we could write these overloads with a decorator, e.g.,
+Our microbenchmark results show that a pure Python implementation of
+``try_array_function_override`` adds approximately 2-4 microseconds of
+overhead to each function call without any overloaded arguments.
+This is acceptable for functions implemented in Python but probably too
+slow for functions written in C. Fortunately, we expect significantly less
+overhead with a C implementation of ``try_array_function_override``.
+
+A more succinct alternative would be to write these overloads with a decorator,
+e.g.,
 
 .. code:: python
 
     @overload_for_array_function(['array'])
     def broadcast_to(array, shape, subok=False):
-        ... # continue with the definition of broadcast_to
+        ...  # continue with the definition of broadcast_to
 
     @overload_for_array_function(['arrays', 'out'])
     def concatenate(arrays, axis=0, out=None):
-        ... # continue with the definition of concatenate
+        ...  # continue with the definition of concatenate
 
 The decorator ``overload_for_array_function`` would be written in terms
-of ``do_array_function_dance``.
+of ``try_array_function_override``.
 
-The downside of this approach would be a loss of introspection capability
-for NumPy functions on Python 2, since this requires the use of
-``inspect.Signature`` (only available on Python 3). However, NumPy won't
-be supporting Python 2 for `very much longer <http://www.numpy.org/neps/nep-0014-dropping-python2.7-proposal.html>`_.
+Implementing it cleanly would require the signature preserving version of
+``functools.wraps`` (only available on Python 3) and the standard libray's
+``inspect`` module. NumPy won't be supporting Python 2 for
+`very much longer <http://www.numpy.org/neps/nep-0014-dropping-python2.7-proposal.html>`_, but a bigger issue is
+performance: ``inspect`` is written in pure Python, so our prototype
+decorator pretty slow, adding about 15 microseonds of overhead.
```

Comment:
I don't think the wrapper would be an issue is we use annotations, since then the wrapper can do all the work while wrapping the function, creating the relevant `try_array_function_override`. It also has the benefit of generically not having to duplicate names in the wrapper.